### ISSUE TITLE
Tcl syntax and double brace issue fails to match braces

### ISSUE THREAD

**Issue by Meshparts** [sublimehq/Packages#1734]
Brace matching fails with Tcl syntax activated when double braces are used in code:

proc Test {args} {
set test {{1 -1 0}}
}

In this example, last brace fails to be matched with corresponding opening brace.

See also disscussion at 
https://forum.sublimetext.com/t/matching-braces-issue/39393/6

**Comment by deathaxe** [comment]
Do double braces have a certain meaning or are the inner braces to be ignored only?

**Comment by Meshparts** [comment]
In Tcl Braces are used to define a string which content will be send to the interpreter without any further evaluation. One can also use braces to define lists with elements separated by empty spaces. The example above defines a nested list with only one sublist. The single sublist contains the three numbers.

**Comment by deathaxe** [comment]
OK, I guess I found a solution to fix your issue.

But ST still highlights the inner list as string because it expects a whitespace after the `{` like in the following example:

Example:

```TCL
set test {{ 1 -1 0 }}
```
Your descriptions sounds like this assumption is not very perfect?

**Comment by Meshparts** [comment]
Thanks for the help. 
Adding white spaces is an workarround indeed. 
Do I get it right? In the next releases it will work also without white spaces?
I wonder if this bug has also other side effects. Seems like the meaning of braces for Tcl is not rely understood currently in ST.

**Comment by Meshparts** [comment]
"Your descriptions sounds like this assumption is not very perfect?" 
Please reformulate. I don't undertande what you are referring to.

**Comment by deathaxe** [comment]
> Please reformulate.

ST highlights all tokens beginning with a `{` and followed by a none-whitespace character as string until the next `}` is matched. This results in `{1 -1 0}` not to be highlighted as list but as string.

This is just how ST tries to distinguish between a normal string and list content. Seems like that assumption needs some refinement?

> Do I get it right? In the next releases it will work also without white spaces?

Depends on whether the related PR is accepted.

> I wonder if this bug has also other side effects.

Yes, #1681 is related. According to some statements from the core devs about TCL, all arguments are basically strings only. ST tries to highlight them best it can but TCL's syntax seems not to be very restrictive, which requires some heuristics to guess how to handle some expressions.

> Seems like the meaning of braces for Tcl is not rely understood currently in ST.

The only thing I can say about it. It's hard to find somebody who has a detailed understanding about a certain syntax and is able to express this knowledge in ST's sophisticated syntax definitions. The syntax packages are open source, which means you need volunteer contributors to manage the amount and complexity of work.

**Comment by Meshparts** [comment]
> ST highlights all tokens beginning with a { and followed by a none-whitespace character as string until the next } is matched. This results in {1 -1 0} not to be highlighted as list but as string.

This is correct. In Tcl a string can become a list just by applying list functinos to it. But in the first place everything between {} can be considered to be a string, which is not further evaluated by Tcl. In contrast everything between "" is also a string but the content between "" will be evaluated first. So if I use variable names in the string, those will be replaced by their values.

The issue here is not the highlighting but the brace matching. Somehow ST looses the count of the braces in this example:

proc Test {args} **{**
set test {{1 -1 0}}
**}**

The braces that are not matched are bold.

**Comment by Meshparts** [comment]
> The braces that are not matched are bold.
You will have to zoom in to see the bold braces.

**Comment by deathaxe** [comment]
It just doesn't respect the rule that braced strings must contain balanced braces. This issue is addressed in  #1751 by highlighting the most inner `{1 -1 0}` as string only.

> But in the first place everything between {} can be considered to be a string

This is ST's problem. It was easy to handle everything as string, but this would look monotonous. Hence ST tries to guess the type of data to perform better highlighting, which requires some heuristics which may fail in special situations.

One such heuristic is to match a brace followed by none-space (e.g. `{nowhitespace }` ) as string without further parsing its content.

Therefore you need to add the "extra" whitespace after a brace to enable the content to be highlighted as list. Don't have an idea how to avoid this.

**Comment by Meshparts** [comment]
> Therefore you need to add the "extra" whitespace after a brace to enable the content to be highlighted as list. Don't have an idea how to avoid this.

I would add an "if" to the rule, to exclude a second brace from the rule above.

**Comment by deathaxe** [comment]
This is what #1751 actually does.

**Comment by Meshparts** [comment]
oh, cool.
Thank you for fixing this!

**Comment by dabro666** [comment]
Hi everyone,
As I found, #1751 is not clear enough (picture below).
![github_tcl_1751](https://user-images.githubusercontent.com/31410405/47636129-5c837180-db68-11e8-9260-8890607324fc.png)
Braces does not identified correctly. "Red numbers" is what should be, "orange numbers" is what we have.

**Comment by FichteFoll** [comment]
so you mean when another `{` occurs within a string, it's likely a nested list? That would at least make the 4th level be recognized correctly.

Are `{third THREE}` and `{fourth FOUR}` strings then?

**Comment by dabro666** [comment]
@FichteFoll 
Yes, you are right, it's nested lists. 

I can't use backslash because it ruins the lists' structure. E.g.: 
```tcl  
% llength {{first ONE} {{second TWO} second TWO} {third THREE {third THREE}} {fourth FOUR}}
4
% llength {{first ONE} \{{second TWO} second TWO\} {third THREE {third THREE}} {fourth FOUR}}
7
```

**Comment by deathaxe** [comment]
Basically everything is a string and nested brace quoted strings must be balanced as well. This was my first approach. But than I forgot about something like `{third {THREE}}` to be possible :-/

**Comment by deathaxe** [comment]
ST handles all braces followed by none-whitespace as strings without parsing any details. This is a heuristic ST used to do some guess work about what the `{` might mean. I can't judge whether it is ok or not, but this change was made not too far in the past.

Changing your initial example as follows would fix nested list highlighting as ST expects a whitespace after `{` to indicate it to be a nested list.

```TCL
foreach var ${list_of_vars} {
    set list_of_lists { { first ONE } { { second TWO } second TWO} { third THREE { third THREE } } { fourth FOUR } }
}
```

**Comment by deathaxe** [comment]
@dabro666: You underlined the following as orange:

![image](https://user-images.githubusercontent.com/16542113/47669506-5367c400-dbab-11e8-837e-9dd6119ffcb8.png)

What are your expectations about it?

Handle the `{second TWO}` as nested list or everything as string? Basically anything is, but ST tries to guess what  it might be.

I start feeling like highlighting something like `{second TWO}` as string is not a good idea at all. Maybe better to force everything to be something like a nested list? Not sure.

**Comment by dabro666** [comment]
@deathaxe 
At my point of view it would be better to identify everything as nested list. But it depends on user preferences and specific of work.
Unfortunately, I can't use whitespaces. For example,"split" return empty elements:
```tcl
% split { one two three }
{} one two three {}
```
So, in some cases this requires additional actions to work with this kind of lists.

**Comment by Meshparts** [comment]
It cannot be a solution to add white spaces aft3er braces. As stated above, this makes a difference in Tcl. Logically a string with white spaces is not the same as a string without white spaces.

**Comment by deathaxe** [comment]
I start digging into the TCL language and ST's syntax at the moment, reading through all the issues of the last years.

From what I've learned so far the major issue is that TCL does not have a strong syntax rule about how to handle the content of such braces. Trying to fix your list highlighting will break other stuff like brace-quoted-regexp patterns. I can't see a syntactical rule to distinguish between a list, a string, a regexp pattern or even nested commands.

In short: TCL looks like a language which is so poor and loose that it can't be highlighted with a static lexer using regexp patterns. It would require dynamic parsing and interpreting the expressions to get an idea about the content.

Example:

What is the following expression? A list or a string?

```
{Hello I am fine with { but not with } at all}
```

You can't give an answer? ST can't as well.

The only solution seems to not highlight that stuff at all.

**Comment by Meshparts** [comment]
> What is the following expression? A list or a string?

The expression is a string.
In Tcl everything is a string, see https://wiki.tcl-lang.org/page/everything+is+a+string

I would handle everything that is between braces AND in only one single line as a string.

Furthermore and apart from the highlighting: The issue here is a brace matching problem. The braces belonging to the "proc" procedure are not matched because something goes wrong in the contained lines. This is a simple brace counting problem that every advanced text editor can/should handle. 

So how is the syntax highlighting related to the brace counting? Shouldn't this be two different things?

**Comment by FichteFoll** [comment]
No, syntax highlighting is strongly related to "brace counting" because it needs to do brace counting in order to highlight properly. I've lost track a bit on this entire situation, but I believe that the initial problem was that braces weren't counted *within* a string, although I'm not sure because I didn't test or verify anything.

In the mean time, the discussion mostly started revolving around when something should be a list and when we can't tell and need to highlight as a string.

(By the way, I hate Tcl for that reason.)

**Comment by deathaxe** [comment]
The main topic of this issue is "brace-balancing". It got screwed up, because I felt like my PR would fix #1681 as well which @dabro666 commented here instead of doing so in #1681 or #1751.

To fix this issue and #1681 I just needed to learn about how to handle the content of braces as well.

**Comment by Bogdan107** [comment]
I fill bug firstly to [forum.sublimetext.com](https://forum.sublimetext.com/t/bug-syntax-tcl-block-of-code-in-one-line-with-other-blocks-highlighted-as-a-value-of-variable-not-as-block-of-code/64494), but duplicate here.

Steps:

Syntax = Tcl;

Color theme = jEdit;

Example 1: code with incorrectly highlighted words
```
33: proc ::_buttontoggleHandler {cbox args } {
34:   # Invert value only by events with tag=“invoke”
35:   if {$args eq “invoke”} {$cbox invert };
36:   $cbox {*}$args;
37: };
```

Example 2: code with correctly highlighted words
```
33: proc ::_buttontoggleHandler { cbox args } {
34:   # Invert value only by events with tag=“invoke”
35:   if {$args eq “invoke”} { $cbox invert };
36:   $cbox {*}$args;
37: };
```

The only difference between examples - in line 35, beetwen the symbol “{” and words “$cbox invert”.

In Example 1, words `$cbox invert` on the line 35, highlights as an value of variable (foreground=pink).
In Example 2, words `$cbox invert` on the line 35, highlights as a normal block of code.

If block of code spaced between the symbol, which indicate the start of new block of code, then Sublime-Text highlights this block of code as block of code - this is correct.
If block of code has no space between the symbol, which indicate the start of new block of code, then Sumblime-Text highlight this block of code as a value of variable - this is incorrect.
